### DISCUSSION TITLE
setValue is clearing unregistered nested values

### DISCUSSION THREAD

**Discussion by Noitidart** [react-hook-form/react-hook-form#5100] [Q&A]
I am using react-hook-form v7.5.0.

I have created a very simplified demo below, and on code sandbox. Nested unregistered fields in default values are cleared when `setValues` is done. Is there a way to retain nested unregistered values?

### Explanation with reproduction

I have a related code and demo (with codesandbox) pasted below.

We see `nested.firstName` and `nested.enabled` are "registered" due to `watch("nested.enabled")` and `<input type="text" {...form.register("nested.firstName")} />`. We see `nested.lastName` is NOT registered. The demo will show this value vanish.

1. Load that snippet, and on first render you see `form.getValues()` gets logged which outputs:

```
  {
    "msg": "render values",
    "values": {
      "nested": {
        "firstName": "foo",
        "lastName": "bar",
        "enabled": false
      }
    }
  }
```

2. Tap the "toggle"
3. We see on next render the log output now shows:

```
  {
    "msg": "render values",
    "values": {
      "nested": {
        "firstName": "foo",
        "enabled": true
      }
    }
  }
```

We see that `nested.lastName` has vanished!

### Question

Does anyone know why this is? And how to make it stay?

### Demo and code

https://codesandbox.io/s/react-hook-form-setvalues-clears-unregistred-default-values-livx0

```
import { useForm } from "react-hook-form";
import { useEffect } from "react";

export default function App() {
  const form = useForm();

  useEffect(() => {
    form.reset({
      nested: {
        firstName: "foo",
        lastName: "bar",
        enabled: false
      }
    });
  }, []);

  console.log(
    JSON.stringify({ msg: "render values", values: form.getValues() }, null, 2)
  );

  const enabled = form.watch("nested.enabled");

  return (
    <>
      <input type="text" {...form.register("nested.firstName")} />

      <button onClick={() => form.setValue("nested.enabled", !enabled)}>
        Set "enabled" to {String(!enabled)}
      </button>
    </>
  );
}

```

**Comment by bluebill1049** [comment]
hey @Noitidart thanks for the detailed question. This is due to input.lastName is missing in the view, react hook form is embed on the input itself to collect their value.  however, the next update should automatically register the missing input for you. so for now you can resolve this issue by simply `register` that missing input.

`register('nested.lastName')`

this will let hook form know there is such input and value to be retained.

**Reply by Noitidart** [reply]
Thanks for your fast reply and your unbelievable support of the community. Yep that fixes it. However I don't want to register that field. I think it's a change in v7 I read this topic - https://github.com/react-hook-form/react-hook-form/discussions/4565#discussioncomment-564533 - I think I agree with the design change. Any other state should be held elsewhere.

Should I close this topic as duplicate?

**Reply by bluebill1049** [reply]
we can leave this open. This PR will address this issue: https://github.com/react-hook-form/react-hook-form/pull/5069

**Reply by Noitidart** [reply]
Thank you @bluebill1049!
